### PR TITLE
amdllpc auto layout desc fixes

### PR DESCRIPTION
**Pull request by trenouf**
1. Avoid duplicate user data entries.
2. Fixed resource and sampler use of combined resource.
3. Cope with variable with binding but no set.

Change-Id: I4233d4bb5961ea2f188ef222dda90ec1d466b887

### CODE REVIEW

File: tool/llpcAutoLayout.cpp [hunk 1]
```diff
@@ -438,7 +444,8 @@ void DoAutoLayoutDesc(
     }
 
     // Collect ResourceMappingNode entries in sets.
-    std::map<uint32_t, std::vector<ResourceMappingNode>> resNodeSets;
+    std::map<uint32_t, ResNodeSet> resNodeSets;
```
Comment (trenouf) [reply tool/llpcAutoLayout.cpp]: This code is only used in standalone amdllpc when compiling something other than a pipeline. In that configuration, in all existing tests, it only compiles a single shader. Therefore I think your suggestions are unnecessary. Can I leave it as it is?

Comment (linqun) [inline tool/llpcAutoLayout.cpp]: 1. please move them to BuildPipeline or declare them as global variable.
We need use same resource node cross shader stage. especially when we compile merged shader (vs + gs) on gfx9. 
we will hit an assert in driver if the resource node is different for same resource in different shader stage 
2. please include stream out node and push constant node in the map too.

File: tool/llpcAutoLayout.cpp [hunk 2]
```diff
@@ -53,6 +53,12 @@ using namespace llvm;
 using namespace Llpc;
 using namespace SPIRV;
 
+struct ResNodeSet
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: ResNodeSet -> ResourceNodeSet

File: tool/llpcAutoLayout.cpp [hunk 3]
```diff
@@ -461,10 +468,25 @@ void DoAutoLayoutDesc(
...
-                    pVar->hasDecorate(DecorationDescriptorSet, 0, &descSet))
+                SPIRVWord descSet = 0;
+                if (pVar->hasDecorate(DecorationBinding, 0, &binding))
                 {
+                    // This copes with a variable having a binding but no set. I saw that happen in
+                    // shaderdb/OpDecorationGroup_TestGroupAndGroupMember_lit.spvas
+                    pVar->hasDecorate(DecorationDescriptorSet, 0, &descSet);
+
+                    // Find/create the node entry for this set and binding.
+                    ResNodeSet& set = resNodeSets[descSet];
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: set -> resNodeSet

File: tool/llpcAutoLayout.cpp [hunk 4]
```diff
@@ -53,6 +53,12 @@ using namespace llvm;
 using namespace Llpc;
 using namespace SPIRV;
 
+struct ResNodeSet
+{
+    std::vector<ResourceMappingNode> nodes;
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: Add comment
nodes // Collections of resource mapping nodes

File: tool/llpcAutoLayout.cpp (hunk 2 above)
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: Add comment

// Represent the info of resource node set.

File: tool/llpcAutoLayout.cpp [hunk 5]
```diff
@@ -473,64 +495,94 @@ void DoAutoLayoutDesc(
...
-                    std::vector<ResourceMappingNode>& set = resNodeSets[descSet];
-                    uint32_t nextOffset = 0;
-                    if (set.size() != 0)
-                    {
-                        nextOffset = set.back().offsetInDwords + set.back().sizeInDwords;
-                    }
-                    set.push_back({});
-
+                    // Map the SPIR-V opcode to descriptor type and size.
+                    ResourceMappingNodeType type;
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: type -> nodeType

File: tool/llpcAutoLayout.cpp [hunk 6]
```diff
@@ -461,10 +468,25 @@ void DoAutoLayoutDesc(
         default:
             {
                 SPIRVWord binding = SPIRVID_INVALID;
-                SPIRVWord descSet = SPIRVID_INVALID;
-                if (pVar->hasDecorate(DecorationBinding, 0, &binding) &&
-                    pVar->hasDecorate(DecorationDescriptorSet, 0, &descSet))
+                SPIRVWord descSet = 0;
+                if (pVar->hasDecorate(DecorationBinding, 0, &binding))
                 {
+                    // This copes with a variable having a binding but no set. I saw that happen in
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: Change the comment:

// NOTE: Test "shaderdb/OpDecorationGroup_TestGroupAndGroupMember_lit.spvas" defines a variable having a binding but no set. We should handle this case.

File: tool/llpcAutoLayout.cpp [hunk 7]
```diff
@@ -461,10 +468,25 @@ void DoAutoLayoutDesc(
...
+                    // shaderdb/OpDecorationGroup_TestGroupAndGroupMember_lit.spvas
+                    pVar->hasDecorate(DecorationDescriptorSet, 0, &descSet);
+
+                    // Find/create the node entry for this set and binding.
+                    ResNodeSet& set = resNodeSets[descSet];
+                    auto iteratorAndCreated = set.bindingMap.insert({binding, set.nodes.size()});
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: iteratorAndCreated  -> it

File: tool/llpcAutoLayout.cpp [hunk 8]
```diff
@@ -53,6 +53,12 @@ using namespace llvm;
 using namespace Llpc;
 using namespace SPIRV;
 
+struct ResNodeSet
+{
+    std::vector<ResourceMappingNode> nodes;
+    std::map<uint32_t, uint32_t> bindingMap;
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: Add comment
bindingMap; // Map from binding ID to set ID

File: tool/llpcAutoLayout.cpp [hunk 9]
```diff
@@ -461,10 +468,25 @@ void DoAutoLayoutDesc(
...
+                if (pVar->hasDecorate(DecorationBinding, 0, &binding))
                 {
+                    // This copes with a variable having a binding but no set. I saw that happen in
+                    // shaderdb/OpDecorationGroup_TestGroupAndGroupMember_lit.spvas
+                    pVar->hasDecorate(DecorationDescriptorSet, 0, &descSet);
+
+                    // Find/create the node entry for this set and binding.
+                    ResNodeSet& set = resNodeSets[descSet];
+                    auto iteratorAndCreated = set.bindingMap.insert({binding, set.nodes.size()});
+                    uint32_t vectorIndex = iteratorAndCreated.first->second;
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: vectorIndex -> nodeIndex

File: tool/llpcAutoLayout.cpp [hunk 10]
```diff
@@ -461,10 +468,25 @@ void DoAutoLayoutDesc(
...
+                    if (iteratorAndCreated.second)
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: iteratorAndCreated.second -> it.second != 0

File: tool/llpcAutoLayout.cpp (hunk 1 above)
Comment (linqun) [reply tool/llpcAutoLayout.cpp]: OK, let's fix it in a separate commit.

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit d018315f1a4c6198386e6ad71b9b96159285bfeb
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev b563d58ad6dc7c2daee9b73801dcc1321686b73b</li>
<li>SPVGEN: dev 5378c7618a2a3bf60ba01a86db980930e8581146</li>
<li>XGL: dev 6aa9683c02873844f932a920bdb4f688dedec724</li>
<li>LLVM: amd-vulkan-dev 1ca248e6fc6d4d15491756f92f4da139222ef3ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 901/901 (100.0%)</li>
<li>Failed: 0/901 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>


</ul></details>

File: tool/llpcAutoLayout.cpp (hunk 7 above)
Comment (trenouf) [reply tool/llpcAutoLayout.cpp]: Disagree -- this is not an iterator, it is a std::pair of an iterator and a flag saying whether insert() created a new entry in the map. So my name makes sense.

File: tool/llpcAutoLayout.cpp (hunk 10 above)
Comment (trenouf) [reply tool/llpcAutoLayout.cpp]: See above.

File: tool/llpcAutoLayout.cpp (hunk 2 above)
Comment (trenouf) [reply tool/llpcAutoLayout.cpp]: Superseded by your comment below on the same line.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a8296a3c63186bb8b741e61b34a54702c2d09160
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev b563d58ad6dc7c2daee9b73801dcc1321686b73b</li>
<li>SPVGEN: dev 5378c7618a2a3bf60ba01a86db980930e8581146</li>
<li>XGL: dev 6aa9683c02873844f932a920bdb4f688dedec724</li>
<li>LLVM: amd-vulkan-dev 1ca248e6fc6d4d15491756f92f4da139222ef3ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 901/901 (100.0%)</li>
<li>Failed: 0/901 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>


</ul></details>

File: tool/llpcAutoLayout.cpp [hunk 11]
```diff
@@ -53,6 +53,12 @@ using namespace llvm;
 using namespace Llpc;
 using namespace SPIRV;
 
+struct ResourceNodeSet
```
Comment (amdrexu) [inline tool/llpcAutoLayout.cpp]: Add the comment for the whole structure

// Represent the info of a collection of resource nodes, including their binding maps.
strcuct ResourceNodeSet

**Pull request merged**